### ISSUE TITLE
slow performance when using Deconvolution for bilinear upsampling

### ISSUE THREAD

**Issue by zym1010** [BVLC/caffe#5398]
### Issue summary

When using Deconvolution Layer to perform bilinear upsampling on feature maps, following instruction in <http://caffe.berkeleyvision.org/doxygen/classcaffe_1_1BilinearFiller.html>, I noticed that this operation is extremely slow, especially when feature map is small, and number of channels is large.

### Steps to reproduce

~~~python
from __future__ import absolute_import, division, print_function
from caffe import layers as L, params as P
import caffe
from tempfile import NamedTemporaryFile
from os import remove
caffe.set_mode_gpu()


# In[2]:

def conv_net(group):
    # check <https://github.com/BVLC/caffe/issues/4052> on how to define Deconvolution layer.
    # input_size is a two dim tuple, and scale_factor is an integer >= 1
    n = caffe.NetSpec()
    n.data = L.Input(shape=[{'dim': list((128,64,14,14))}])
    conv_param = {
        'kernel_size':4,
        'num_output':64,
        'group':group,
        'stride': 2,
        'pad': 1,
        'weight_filler': dict(type='xavier')
    }
    n.conv1 = L.Deconvolution(n.data, convolution_param=conv_param)
    return str(n.to_proto())

def get_caffe_conv_net(group):
    with NamedTemporaryFile(delete=False) as f_temp_file:
        f_temp_file.write(conv_net(group))
    create_net = caffe.Net(f_temp_file.name, caffe.TEST)
    remove(f_temp_file.name)
    
    for layer_name, blob in create_net.blobs.iteritems():
        print(layer_name + '\t' + str(blob.data.shape))
        
    for param_name, param in create_net.params.iteritems():
        print(param_name + '\t' + str(param[0].data.shape))
    
    return create_net


# In[3]:

# seems that net2 is much slower than net1 when doing deconv.
net1 = get_caffe_conv_net(1)
net2 = get_caffe_conv_net(64)

get_ipython().magic(u'timeit -n 10 net1.forward()')
get_ipython().magic(u'timeit -n 10 net2.forward()')
~~~

the result is 
~~~
data	(128, 64, 14, 14)
conv1	(128, 64, 28, 28)
conv1	(64, 64, 4, 4)
data	(128, 64, 14, 14)
conv1	(128, 64, 28, 28)
conv1	(64, 1, 4, 4)
10 loops, best of 3: 15.3 ms per loop
10 loops, best of 3: 46.9 ms per loop
~~~

notice that net2 is much slower than net1, though with much fewer parameters.

### Your system configuration
Operating system: CentOS 6
Compiler: GCC 4.9.1
CUDA version (if applicable): 8.0.61
CUDNN version (if applicable): 5.1.10
BLAS: OpenBLAS 0.2.19
Python or MATLAB version (for pycaffe and matcaffe respectively): 2.7.12

**Comment by shelhamer** [comment]
I have a vague memory of this but thought it had been fixed. What happens if you set `group` to 1? It might scale poorly with the number of groups.

**Comment by zym1010** [comment]
@shelhamer as shown in my test results, `group`=1 works better than `group`=64. However, the bilinear filler only works in the latter case.

**Comment by shelhamer** [comment]
Ah, I misread your test—thanks for following up. This is what I though then: there is slowdown linear in the number of groups. See the loops for group here: https://github.com/BVLC/caffe/blob/master/src/caffe/layers/base_conv_layer.cpp

> However, the bilinear filler only works in the latter case [...of num. output == group]

Not exactly: if you only set nonzero weights on the "diagonal" of output-input channels you can do per-channel interpolation by trading space for time. My net surgery helper for interpolation handles |groups| == 1 or |groups| == |channels|.

Note that there might be an alternative implementation through cuDNN. I'm closing this is as known behavior.

**Comment by zym1010** [comment]
@shelhamer 

> Not exactly: if you only set nonzero weights on the "diagonal" of output-input

So what I mean is that, by using group = 1 for better performance. I need to essentially initialize the weights myself, using Python, or something else, instead of staying with the command line tools. I found this to be a little inconvenient. That's all

> Note that there might be an alternative implementation through cuDNN. I'm closing this is as known behavior.

What I showed you is the cuDNN performance. On CPU, I bet the difference won't be that big. I noticed that in cuDNN implementation, there's a async loop over groups as well. No idea why such async loop will make such slowdown.
